packages/jest-config/src/normalize.js

```diff
@@ -268,32 +276,36 @@ const normalizeReporters = (options: InitialOptions, basedir) => {
   return options;
 };
 
-function normalize(options: InitialOptions, argv: Object = {}) {
+function normalize(options: InitialOptions, argv: Argv) {
   const {hasDeprecationWarnings} = validate(options, {
     comment: DOCUMENTATION_NOTE,
     deprecatedConfig: DEPRECATED_CONFIG,
     exampleConfig: VALID_CONFIG,
   });
 
-  options = setFromArgv(options, argv);
-  normalizeReporters(options);
-  normalizePreprocessor(options);
-  normalizeRootDir(options);
-  normalizeMissingOptions(options);
+  options = normalizePreprocessor(
+    normalizeReporters(
+      normalizeMissingOptions(normalizeRootDir(setFromArgv(options, argv))),
+    ),
+  );
 
   if (options.preset) {
     options = setupPreset(options, options.preset);
   }
+
   if (options.testEnvironment) {
     options.testEnvironment = getTestEnvironment(options);
   }
+
   if (!options.roots && options.testPathDirs) {
     options.roots = options.testPathDirs;
+    delete options.testPathDirs;
   }
 
   const babelJest = setupBabelJest(options);
   const newOptions = Object.assign({}, DEFAULT_CONFIG);
-
+  // Cast back to exact type
+  options = (options: InitialOptions);
   Object.keys(options).reduce((newOptions, key) => {
```

Comment:
we could avoid these null checks if we filter `Object.keys(options)` first, if Flow understands that.
```js
Object.keys(options)
  .filter(key => options[key] !== undefined)
  .reduce((newOptions, key) => 
```